modules/player/src/main/java/unnamed/mmo/player/PlayerImpl.java

```diff
@@ -122,4 +130,33 @@ private void updateDiggingBlock() {
         var packet = new BlockBreakAnimationPacket(getEntityId() + 1, diggingBlock, stage);
         sendPacket(packet);
     }
+
+    /**
+     * Adds a new permanent modifier to this player
+     * @param modifierType The type of modifier to modify, such as dig speed, melee damage, etc. This modifier type should be in the modifier registry, and can be checked with {@link ModifierType#doesModifierExist(String)}
+     * @param modifierId The id of this modifier. If a player already has this modifier id, it will be overridden. This id can be basically any string, and it should be unique to each source so modifiers can stack properly
+     * @param amount The amount to modify by
+     * @param operation The operation with which to modify by
+     */
+    public void addPermanentModifier(String modifierType, String modifierId, double amount, ModifierOperation operation) {
+        if (ModifierType.doesModifierExist(modifierType)) {
+            currentModifiers.putIfAbsent(modifierType, new ModifierList(ModifierType.getBaseValue(modifierType)));
```

Comment:
This might be the most nitpicky thing I have ever commented about, but you can do
```java
currentModifiers.computeIfAbsent(modifierType, type -> new ModifierList(ModifierType.getBaseValue(type))
        .addPermanentModifier(modifierId, amount, operation);
```